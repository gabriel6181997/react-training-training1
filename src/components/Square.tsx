import { FC } from "react";
import { SquareProps } from "../type";

export const Square: FC<SquareProps> = (props) => {
  let className = "square";
  if (props.isHighlighted) {
    className += " highlight";
  }

  return (
    <button className={className} onClick={props.onClick}>
      {props.value}
    </button>
  );
};
