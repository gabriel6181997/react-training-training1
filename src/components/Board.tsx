import { FC } from "react";
import { BoardProps } from "../type";
import { Square } from "./Square";

export const Board: FC<BoardProps> = (props) => {
  const renderSquare = (i: number, isHighlighted: boolean) => {
    return (
      <Square
        isHighlighted={isHighlighted}
        key={i}
        value={props.squares[i]}
        onClick={() => props.onClick(i)}
      />
    );
  };

  return (
    <div>
      {Array(3)
        .fill(0)
        .map((row, i) => {
          return (
            <div className="board-row" key={i}>
              {Array(3)
                .fill(0)
                .map((col, j) => {
                  const index: number = i * 3 + j;
                  const highlight =
                    props.winLine && props.winLine.indexOf(index) !== -1;
                  if (highlight) {
                    return renderSquare(index, highlight);
                  } else {
                    return renderSquare(index, false);
                  }
                })}
            </div>
          );
        })}
    </div>
  );
};
