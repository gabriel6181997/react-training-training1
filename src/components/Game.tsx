import { useGame } from "../hooks/useGame";
import { Board } from "./Board";
import "./../index.css";

export const Game = () => {
  const { status, moves, current, toggleAsc, handleClick, settlement, isAsc } =
    useGame();

  return (
    <div className="game">
      <div className="game-board">
        <Board
          squares={current.squares}
          onClick={(i: number) => handleClick(i)}
          winLine={settlement ? settlement.line : []}
        />
      </div>
      <div className="game-info">
        <div>{status}</div>
        <div>
          <button onClick={() => toggleAsc()}>
            {isAsc ? "Sort in descending order" : "Sort in ascending order"}
          </button>
        </div>
        <ol>{isAsc ? moves : moves.reverse()}</ol>
      </div>
    </div>
  );
};
