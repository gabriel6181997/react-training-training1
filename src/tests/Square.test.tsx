import renderer from "react-test-renderer";
import { Square } from "../components/Square";
import { expect, test } from "@jest/globals";

test("add className highlight when isHighlighted is true", () => {
  const tree = renderer
    .create(<Square isHighlighted={true} value={null} onClick={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("not add className highlight when isHighlighted is false", () => {
  const tree = renderer
    .create(<Square isHighlighted={false} value={null} onClick={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("show X when value is X", () => {
  const tree = renderer
    .create(<Square isHighlighted={false} value={"X"} onClick={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("show O when the value is O", () => {
  const tree = renderer
    .create(<Square isHighlighted={false} value={"O"} onClick={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("show nothing when the value is null", () => {
  const tree = renderer
    .create(<Square isHighlighted={false} value={null} onClick={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// test("call onClick when the square is clicked", () => {
//   const onClickSquare = jest.fn();
//   const tree = renderer
//     .create(
//       <Square isHighlighted={false} value={null} onClick={onClickSquare} />
//     )
//     .toJSON();
//   if ("tree" in renderer.ReactTestRenderer) {
//     tree?.props.onClick();
//   }
//   expect(onClickSquare).toHaveBeenCalled();
// });
