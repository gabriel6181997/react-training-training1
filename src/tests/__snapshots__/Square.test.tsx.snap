// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add className highlight when isHighlighted is true 1`] = `
<button
  className="square highlight"
  onClick={[Function]}
/>
`;

exports[`not add className highlight when isHighlighted is false 1`] = `
<button
  className="square"
  onClick={[Function]}
/>
`;

exports[`show O when the value is O 1`] = `
<button
  className="square"
  onClick={[Function]}
>
  O
</button>
`;

exports[`show X when value is X 1`] = `
<button
  className="square"
  onClick={[Function]}
>
  X
</button>
`;

exports[`show nothing when the value is null 1`] = `
<button
  className="square"
  onClick={[Function]}
/>
`;
